src/search.rs

```diff
@@ -0,0 +1,146 @@
+use crate::search_error::SearchError;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+pub fn search(directory: &Path) -> Result<PathBuf, SearchError> {
+  let mut files = Vec::new();
+  let dir = fs::read_dir(directory).map_err(|io_error| SearchError::Io {
+    io_error,
+    directory: directory.to_owned(),
+  })?;
+  for entry in dir {
+    let entry = entry.map_err(|io_error| SearchError::Io {
+      io_error,
+      directory: directory.to_owned(),
+    })?;
+    if let Some(name) = entry.file_name().to_str() {
+      use caseless::default_caseless_match_str;
+      if default_caseless_match_str(name, "justfile") {
+        files.push(entry.path());
+      }
+    }
+  }
+  if files.len() == 1 {
+    Ok(files.pop().unwrap())
+  } else if files.len() > 1 {
+    Err(SearchError::MultipleCandidates { candidates: files })
```

Comment:
If you rename `files` to `candidates`, then you can do: `SearchError::MultipleCandidates { candidates }`.